rust/operator/src/kafka_controller.rs

```diff
@@ -0,0 +1,560 @@
+pub enum Error {
+    #[snafu(display("object has no namespace"))]
+    ObjectHasNoNamespace,
+    #[snafu(display("object defines no version"))]
+    ObjectHasNoVersion,
+    #[snafu(display("object defines no broker role"))]
+    NoBrokerRole,
+    #[snafu(display("failed to calculate global service name"))]
+    GlobalServiceNameNotFound,
+    #[snafu(display("failed to calculate service name for role {}", rolegroup))]
+    RoleGroupServiceNameNotFound {
+        rolegroup: RoleGroupRef<KafkaCluster>,
+    },
+    #[snafu(display("failed to apply global Service"))]
+    ApplyRoleService {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to apply Service for {}", rolegroup))]
+    ApplyRoleGroupService {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<KafkaCluster>,
+    },
+    #[snafu(display("failed to build ConfigMap for {}", rolegroup))]
+    BuildRoleGroupConfig {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<KafkaCluster>,
+    },
+    #[snafu(display("failed to apply ConfigMap for {}", rolegroup))]
+    ApplyRoleGroupConfig {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<KafkaCluster>,
+    },
+    #[snafu(display("failed to apply StatefulSet for {}", rolegroup))]
+    ApplyRoleGroupStatefulSet {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<KafkaCluster>,
+    },
+    #[snafu(display("failed to generate product config"))]
+    GenerateProductConfig {
+        source: stackable_operator::product_config_utils::ConfigError,
+    },
+    #[snafu(display("invalid product config"))]
+    InvalidProductConfig {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to serialize zoo.cfg for {}", rolegroup))]
+    SerializeZooCfg {
+        source: stackable_operator::product_config::writer::PropertiesWriterError,
+        rolegroup: RoleGroupRef<KafkaCluster>,
+    },
+    #[snafu(display("object is missing metadata to build owner reference"))]
+    ObjectMissingMetadataForOwnerRef {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to build discovery ConfigMap"))]
+    BuildDiscoveryConfig {
+        source: discovery::Error,
+    },
+    #[snafu(display("failed to apply discovery ConfigMap"))]
+    ApplyDiscoveryConfig {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to update status"))]
+    ApplyStatus {
+        source: stackable_operator::error::Error,
+    },
+    RoleGroupNotFound,
+}
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub async fn reconcile_kafka(kafka: KafkaCluster, ctx: Context<Ctx>) -> Result<ReconcilerAction> {
+    tracing::info!("Starting reconcile");
+    let client = &ctx.get_ref().client;
+
+    let validated_config = validate_all_roles_and_groups_config(
+        &kafka.spec.version,
```

Comment:
I would still pack that via `kafka_version` in case we do some checks etc.

Reply:
Good point.